stan/math/rev/functor/algebra_solver_powell.hpp

```diff
@@ -15,168 +18,117 @@
-  using Fx = hybrj_functor_solver<Fs, F, double, double>;
-  Fx fx(Fs(), f, x_val, y, dat, dat_int, msgs);
-  Eigen::HybridNonLinearSolver<Fx> solver(fx);
+  check_nonzero_size("algebra_solver_powell", "initial guess", x_val);
+  check_finite("algebra_solver_powell", "initial guess", x_val);
+  check_nonnegative("alegbra_solver_powell", "relative_tolerance",
+                    relative_tolerance);
+  check_nonnegative("algebra_solver_powell", "function_tolerance",
+                    function_tolerance);
+  check_positive("algebra_solver_powell", "max_num_steps", max_num_steps);
+
+  // Construct the Powell solver
+  auto myfunc = [&](const auto& x) {
+    return apply([&](const auto&... args) { return f(x, msgs, args...); },
+                 args_vals_tuple);
+  };
+
+  hybrj_functor_solver<decltype(myfunc)> fx(myfunc);
+  Eigen::HybridNonLinearSolver<decltype(fx)> solver(fx);
 
   // Check dimension unknowns equals dimension of system output
   check_matching_sizes("algebra_solver", "the algebraic system's output",
                        fx.get_value(x_val), "the vector of unknowns, x,", x);
 
-  // Compute theta_dbl
-  Eigen::VectorXd theta_dbl = x_val;
-  solver.parameters.xtol = relative_tolerance;
-  solver.parameters.maxfev = max_num_steps;
-  solver.solve(theta_dbl);
+  // Solve the system
+  Eigen::VectorXd theta_dbl
+      = algebra_solver_powell_(solver, fx, x_val, 0, relative_tolerance,
+                               function_tolerance, max_num_steps);
 
-  // Check if the max number of steps has been exceeded
-  if (solver.nfev >= max_num_steps) {
-    throw_domain_error("algebra_solver", "maximum number of iterations",
-                       max_num_steps, "(", ") was exceeded in the solve.");
-  }
+  Eigen::MatrixXd Jf_x;
+  Eigen::VectorXd f_x;
 
-  // Check solution is a root
-  double system_norm = fx.get_value(theta_dbl).stableNorm();
-  if (system_norm > function_tolerance) {
-    std::ostringstream message;
-    message << "the norm of the algebraic function is " << system_norm
-            << " but should be lower than the function "
-            << "tolerance:";
-    throw_domain_error("algebra_solver", message.str().c_str(),
-                       function_tolerance, "",
-                       ". Consider decreasing the relative tolerance and "
-                       "increasing max_num_steps.");
-  }
+  jacobian(myfunc, theta_dbl, f_x, Jf_x);
 
-  return theta_dbl;
-}
+  using ret_type = Eigen::Matrix<var, Eigen::Dynamic, -1>;
+  auto arena_Jf_x = to_arena(Jf_x);
+
+  arena_t<ret_type> ret = theta_dbl;
+
+  reverse_pass_callback([f, ret, arena_args_tuple, arena_Jf_x, msgs]() mutable {
+    using Eigen::Dynamic;
+    using Eigen::Matrix;
+    using Eigen::MatrixXd;
+    using Eigen::VectorXd;
+
+    // Contract specificities with inverse Jacobian of f with respect to x.
+    VectorXd ret_adj = ret.adj();
+    VectorXd eta = -arena_Jf_x.transpose().fullPivLu().solve(ret_adj);
```

Comment:
[optional] We can do the LU outside of the reverse pass and save the result if we think this should be fast for multiple reverse passes (see [here](https://github.com/stan-dev/math/blob/develop/stan/math/rev/fun/mdivide_left.hpp#L45) for example) and can do the same for the newton solver.

Reply:
I would like to do this (see my conversation with @SteveBronder below), but I'm running into some strange polymorphism issues. In particular, if I try to exactly recreate the linked example as follows:
```c++
  using ret_type = Eigen::Matrix<var, Eigen::Dynamic, -1>;
  auto arena_Jf_x_ptr = make_chainable_ptr(Jf_x.transpose().lu());

  arena_t<ret_type> ret = theta_dbl;

  reverse_pass_callback([f, ret, arena_args_tuple, arena_Jf_x_ptr, msgs]() mutable {
    using Eigen::Dynamic;
    using Eigen::Matrix;
    using Eigen::MatrixXd;
    using Eigen::VectorXd;

    // Contract specificities with inverse Jacobian of f with respect to x.
    VectorXd ret_adj = ret.adj();
    VectorXd eta = -arena_Jf_x_ptr->solve(ret_adj);
```
I get the following compiler error:
```
In file included from test/unit/math/rev/functor/algebra_solver_test.cpp:2:
./stan/math/rev/functor/algebra_solver_powell.hpp:151:37: error: no member named 'solve' in 'Eigen::Matrix<double, -1, -1, 1, -1, -1>'
    VectorXd eta = -arena_Jf_x_ptr->solve(ret_adj);
                    ~~~~~~~~~~~~~~  ^
```
It seems that, for whatever reason, the pointer is getting coerced to a pointer to the base matrix type instead of the derived LU decomposed matrix type. I've tried a few things to prevent this, but without success. Am I missing something super obvious? After looking at the code for `chainable_object` I don't see why this would happen here but not in the example you linked.